File: transform-hyphens.py
from itertools import combinations
from Levenshtein import distance

def get_hyphen_points(hyphenated):
    return [i for i, char in enumerate(hyphenated) if char == '-']

def insert_hyphens(word, positions):
    result = list(word)
    for pos in sorted(positions, reverse=True):
        result.insert(pos, '-')
    return ''.join(result)

def transform(hyphenated: str, target: str):
    origin = hyphenated.replace("-", "")
    origin_hyphen_points = get_hyphen_points(hyphenated)
    num_hyphens = len(origin_hyphen_points)
    
    possible_positions = range(1, len(target))
    best_result = None
    min_distance = float('inf')

    for hyphen_positions in combinations(possible_positions, num_hyphens):
        candidate = insert_hyphens(target, hyphen_positions)
        current_distance = distance(hyphenated, candidate)
        
        if current_distance < min_distance:
            min_distance = current_distance
            best_result = candidate

    return best_result

# Test cases
test_cases = [
    ("abc-d-ef", "axbxcdxef"),
    ("ne-boj-sa", "nebojsa"),
    ("pret-hod-ny", "predhodny"),
    ("roz-šou-stat", "rˈosʃoʊstat"),
    ("ne-roz-hod-nost", "nˈeroshˌodnost")

]

for hyphenated, target in test_cases:
    result = transform(hyphenated, target)
    print(f"Input: {hyphenated} -> {target}")
    print(f"Output: {result}\n")

File: wlh2ipawlh/src/transform_hyphens.rs
use triple_accel::levenshtein::levenshtein_simd_k_str;
use itertools::Itertools;

fn get_hyphen_points(hyphenated: &str) -> Vec<usize> {
    hyphenated.char_indices()
        .filter_map(|(i, c)| if c == '-' { Some(i) } else { None })
        .collect()
}

fn insert_hyphens(word: &str, positions: &[usize]) -> String {
    let mut result: Vec<char> = word.chars().collect();
    let mut offset = 0;
    for &pos in positions.iter() {
        if pos + offset <= result.len() {
            result.insert(pos + offset, '-');
            offset += 1;
        }
    }
    result.into_iter().collect()
}

pub fn transform(hyphenated: &str, target: &str) -> String {
    let num_hyphens = get_hyphen_points(hyphenated).len();
    let possible_positions: Vec<usize> = (1..target.len()).collect();
    let mut best_distance = hyphenated.len() as u32;
    let mut best_candidate = String::new();

    for hyphen_positions in possible_positions.into_iter().combinations(num_hyphens) {
        if hyphen_positions.first() == Some(&0) || hyphen_positions.last() == Some(&(target.len() - 1)) {
            continue;
        }

        let candidate = insert_hyphens(target, &hyphen_positions);
        if let Some(distance) = levenshtein_simd_k_str(hyphenated, &candidate, best_distance) {
            if distance < best_distance {
                best_distance = distance;
                best_candidate = candidate;
            }
        }
    }

    if best_candidate.is_empty() { target.to_string() } else { best_candidate }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_transform() {
        let test_cases = vec![
            ("abc-d-ef", "axbxcdxef"),
            ("ne-boj-sa", "nebojsa"),
            ("pret-hod-ny", "predhodny"),
            ("roz-šou-stat", "rˈosʃoʊstat"),
            ("ne-roz-hod-nost", "nˈeroshˌodnost"),
        ];

        for (hyphenated, target) in test_cases {
            let result = transform(hyphenated, target);
            println!("Input: {} -> {}", hyphenated, target);
            println!("Output: {}\n", result);
        }
    }
}
